docs/GDPs/gdp-duck-storage.rst

```diff
@@ -0,0 +1,456 @@
+
+:code:`__deepcopy__` and :code:`copy` methods
+   allocate new buffers and copy the contents
+
+:code:`__getitem__`
+   dimensions, for which a certain index is selected are returned as masked, while slices do not reduce dimensionality.
+
+:code:`__setitem__`
+   :ref:`broadcasting: and device selection is equivalent to that of a unary ufunc with a provided output buffer.
+   For example, :code:`stor_out[:,3:5, 0] = stor2d` would be equivalent to
+   :code:`np.positive(stor2d, out=stor_out[:,3:5, 0]`)
+
+:code:`copy`
+   allocate new buffers and copy the contents
+
+The following methods are used to ensure one-sided modifications to CPU or GPU buffers of the
+`SoftwareManagedGPUStorage` are tracked properly. They are no-ops for all other storage classes, but are there so that
+user code can be backend-agnostic in these cases.
+
+The use of these methods should only be necessary, if a reference to the storage buffers is kept and modified outside
+of GT4Py.
+
+:code:`set_device_modified`, :code:`set_host_modified`, :code:`set_device_synchronized`
+   mark a buffer as modified, so that it can be synchronized before the respective other buffer is accessed.
+
+:code:`host_to_device` (:code:`device_to_host`)
+   Triggers a copy from host (device) buffer to the sibling in device (host) memory, if the host (device) is marked as
+   modified or the method is called with `force=True`. After a call to either of these methods, the buffers are flagged
+   as synchronized.
+
+:code:`synchronize`
+   Triggers a copy between host and device buffers if the host or device, respectively are marked as modified. The
+   buffers are marked as in sync as a consequence.
+
+
+Universal Functions
+^^^^^^^^^^^^^^^^^^^
+
+Universal functions, such as mathematical binary operations and logical operators are supported through the
+:code:`numpy.lib.mixins.NDArrayOperatorsMixin` type and the `__array_ufunc__` interface. We support the methods
+`__call__` and `reduce` of the numpy ufunc mechanism.
+
+If the :code:`reduce` method of ufuncs is used, this results in a Storage with the dimensions masked along which the
+reduction was performed.
+
+.. _broadcasting:
+
+Broadcasting
+============
+
+With the term "broadcasting", NumPy describes the ways that different shapes are combined in assignments and
+mathematical operations. We override the default NumPy behavior so that fields are broadcast along the same spatial
+dimension. I.e. adding an :code:`IJ` field :code:`A` of shape :code:`(2, 3)` with a :code:`K` field :code:`B` of shape
+:code:`(4,)` will result in an :code:`IJK` field :code:`C` of shape :code:`(2, 3, 4)`, with `C[i,j,k] = A[i,j]+B[k]`.
+
+Similarly, fields of lower dimension are assigned to such of higher dimension by broadcasting along the missing
+dimensions.
+
+Further, the output buffer can have higher dimensionality than the determined broadcast shape. In this case, the result
+is replicated along the missing dimensions.
+
+
+
+Output Storage Parameters
```

Comment:
Here I'm lost again... Could you explain what these parameters are and where they are?

Reply:
if you have a binary operation such as an addition, which is not performed in-place on an existing buffer, and no output buffer is provided, this will trigger an allocation, resulting in a new storage object. this paragraph describes how the parameters are chosen for the allocated new buffer.